12-offer-encoding.md

```diff
@@ -0,0 +1,846 @@
+
+# Invoices
+
+Invoices are a payment request, and when the payment is made, 
+it can be combined with the invoice to form a cryptographic receipt.
+
+The human-readable prefix for invoices is `lni`.  The recipient can send it in
+response to an `invoice_request` using the `onion_message` `invoice` field
+(in which case, it duplicates the fields including the compulsory
+`invreq_payer_id`), or can be produced spontanously (in which case, it has
+more limited fields, and in particular has no `invreq_payer_id`).
+
+1. `tlv_stream`: `invoice`
+2. types:
+    1. type: 0 (`invreq_metadata`)
+    2. data:
+        * [`...*byte`:`blob`]
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`offer_issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 20 (`offer_quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 22 (`offer_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 80 (`invreq_chain`)
+    2. data:
+        * [`chain_hash`:`chain`]
+    1. type: 82 (`invreq_amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 84 (`invreq_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 86 (`invreq_quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 88 (`invreq_payer_id`)
+    2. data:
+        * [`point`:`key`]
+    1. type: 89 (`invreq_payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 160 (`invoice_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 162 (`invoice_blindedpay`)
+    2. data:
+        * [`...*blinded_payinfo`:`payinfo`]
+    1. type: 164 (`invoice_created_at`)
+    2. data:
+        * [`tu64`:`timestamp`]
+    1. type: 166 (`invoice_relative_expiry`)
+    2. data:
+        * [`tu32`:`seconds_from_creation`]
+    1. type: 168 (`invoice_payment_hash`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 170 (`invoice_amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 172 (`invoice_fallbacks`)
+    2. data:
+        * [`...*fallback_address`:`fallbacks`]
+    1. type: 174 (`invoice_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 176 (`invoice_code`)
+    2. data:
+        * [`...*utf8`:`code`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_payinfo`
```

Comment:
We should also add the route absolute expiry here: [`u32`:`route_expiry`].
I'm not sure whether we want to make that optional somehow, or if it should always be provided.

Mobile wallets with a long-lived connection to an LSP could potentially use payment blinded paths that don't expire without losing privacy (since it's all unannounced "edge" channels anyway), but it's probably better to have one even in that case, to force developers to think about the fact that payment blinded paths eventually expire and that needs to be handled.

Reply:
There is already `invoice_relative_expiry`. The route expiry should be the same, there is no reason to make it longer as the route is specific to this invoice. Do you have in mind an extension of the protocol where the payment route would be reused later?

Reply:
Got it, using `invoice_relative_expiry` sounds good, but we should explicitly mention that any blinded paths included in the invoice should expire at that time, this is currently missing from the spec and useful for implementers.

Reply:
Wait, `invoice_relative_expiry` is a duration in seconds, we want an expiry at a specific block height for blinded paths, so something needs to be updated here.

Reply:
I don't understand why you want to know the expiry for the route, what use case do you imagine? The expiry for the route should be chosen so that it is rare that it expires before the invoice but not too long after the invoice.
I actually think that the invoice request already demonstrates readiness to pay, so the invoice should be paid just seconds after being received.

Reply:
This is linked to the `min_final_expiry_delta` issue that is currently being discussed on the eclair offers PR.

When sending the payment, the payer needs to use a `cltv_expiry` for the recipient that is greater than a few blocks from the current expiry to avoid force-closing. We need to make sure that this doesn't lead to a violation of the blinded path expiry, so the recipient needs to take that into account but doesn't know beforehand what value the payer will use.

I agree that **if the recipient knows an upper bound on what the payer will use**, then we're fine, because it can choose the blinded path expiry based on the `invoice_relative_expiry`, the expected block frequency and the upper bound on the payer cltv expiry value. But to make sure that works, we need to at least detail that in the specification, and highlight that senders MUST NOT add some final cltv expiry randomization like we currently do for Bolt 11 invoices (it's unnecessary with blinded paths because the recipient can handle its own privacy by adding dummy hops), otherwise that upper bound can easily be exceeded and payments will fail.

Reply:
question to this related topic:

Do I understand it correctly that in case of Bolt12 invoices combined with blinded paths, the sender will set the outgoing_cltv_value for the final hop because he constructs the route. However we need to require that the receiver includes a sufficient "final_delta/dummy_hop" so he can be sure in an edge case where the `sender_current_height` != `current_height_when_receiving` are different (blocks are found between sending an receiving the payment) that the real cltv_lock is way greater than the `outgoing_cltv_value` so that the receiver does not run into a force close or reject the payment at all?

Will nodes reject every forwarding/htlc if the `outgoing_cltv_value` is already behind the blockchain tip?

Or should we even require the sender to set a reasonable `buffer` value so that "current height when sending" and "current height when receiving" are guaranteed to be after the current blockchain tip for both?

Resource: https://gist.github.com/t-bast/b1371d357a2c5f3e8c09514a62db7079